### PR TITLE
adds homology classifier

### PR DESCRIPTION
**Pull request by anilthanki**

### CODE REVIEW

File: tools/ete/homology_classifier.xml [hunk 1]
```diff
@@ -0,0 +1,100 @@
+<tool id="homology_classifier" name="Homology Classifier and Filter" version="@VERSION@.1">
```
Comment (nsoranzo) [inline tools/ete/homology_classifier.xml]: Can you prepend the file and tool names with `ete_` for consistency? Also the Python one.
And you can remove the `.1` from the version.

File: tools/ete/homology_classifier.xml [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    <description>from a genetree utilising the ETE Toolkit</description>
+    <macros>
+        <import>ete_macros.xml</import>
+    </macros>
+    <expand macro="requirements" />
+    <stdio>
+        <!-- Anything other than zero is an error -->
+        <exit_code range="1:" />
+    </stdio>
+    <command>
```
Comment (nsoranzo) [inline tools/ete/homology_classifier.xml]: You can add ` detect_errors="exit_code"` to `<command>` and remove the `<stdio>` section.

File: tools/ete/homology_classifier.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    species_dict = {}
+
+    for val in species_list:
+        count = "one"
+        if val in species_dict:
+            count = "many"
+        species_dict[val] = count
+
+    homologis = {
```
Comment (nsoranzo) [inline tools/ete/homology_classifier.py]: homologies ?

File: tools/ete/homology_classifier.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        'one-to-one': [],
+        'one-to-many': [],
+        'many-to-one': [],
+        'many-to-many': [],
+        'paralogs': []
+    }
+
+    # stores relevant homolgy types in dict
```
Comment (nsoranzo) [inline tools/ete/homology_classifier.py]: s/homolgy/homology/

File: tools/ete/homology_classifier.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        if '_' in leaves_list[i]:
+            species_list.append(leaves_list[i].split("_")[1])
+        else:
+            gene_lose_list.append(i)
+
+    leaves_list = [i for j, i in enumerate(leaves_list) if j not in gene_lose_list]
+
+    species_dict = {}
+
+    for val in species_list:
```
Comment (nsoranzo) [inline tools/ete/homology_classifier.py]: A better name for `val`, `species` maybe?

File: tools/ete/homology_classifier.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    species_list = []
+    gene_lose_list = []
+
+    for i, val in enumerate(leaves_list):
+        if '_' in leaves_list[i]:
+            species_list.append(leaves_list[i].split("_")[1])
+        else:
+            gene_lose_list.append(i)
+
+    leaves_list = [i for j, i in enumerate(leaves_list) if j not in gene_lose_list]
```
Comment (nsoranzo) [inline tools/ete/homology_classifier.py]: I'd do this filtering before the above `for` cycle, i.e.:
```python
leaves_list = [_ for _ in leaves_list if '_' in _]
for leaf in leaves_list:
    species_list.append(leaf.split("_")[1])
```

No need for `gene_lose_list`.

File: tools/ete/homology_classifier.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+    options, args = parser.parse_args()
+
+    if options.genetree is None:
+        parser.error("--genetree option must be specified, GeneTree in nhx format")
+
+    # reads single gene tree
+    genetree = PhyloTree(options.genetree)
+
+    leaves_list = genetree.get_leaf_names()
```
Comment (nsoranzo) [inline tools/ete/homology_classifier.py]: Can you add a comment with the expected format of `leaves_list` elements?

File: tools/ete/homology_classifier.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    homologis = {
+        'one-to-one': [],
+        'one-to-many': [],
+        'many-to-one': [],
+        'many-to-many': [],
+        'paralogs': []
+    }
+
+    # stores relevant homolgy types in dict
+    for i, val in enumerate(leaves_list):
```
Comment (nsoranzo) [inline tools/ete/homology_classifier.py]: s/val/leaf1/

File: tools/ete/homology_classifier.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        for j in range(i + 1, len(leaves_list)):
+            id1 = leaves_list[i].split(":")[1] if ":" in leaves_list[i] else leaves_list[i]
```
Comment (nsoranzo) [inline tools/ete/homology_classifier.py]: s/leaves_list[i]/leaf1/ (3 times)

File: tools/ete/homology_classifier.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        'one-to-one': [],
+        'one-to-many': [],
+        'many-to-one': [],
+        'many-to-many': [],
+        'paralogs': []
+    }
+
+    # stores relevant homolgy types in dict
+    for i, val in enumerate(leaves_list):
+        for j in range(i + 1, len(leaves_list)):
```
Comment (nsoranzo) [inline tools/ete/homology_classifier.py]: `for leaf2 in leaves_list[i + 1, ]:`

File: tools/ete/homology_classifier.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+            id1 = leaves_list[i].split(":")[1] if ":" in leaves_list[i] else leaves_list[i]
+            id2 = leaves_list[j].split(":")[1] if ":" in leaves_list[j] else leaves_list[j]
```
Comment (nsoranzo) [inline tools/ete/homology_classifier.py]: s/leaves_list[j]/leaf2/ (3 times)

File: tools/ete/homology_classifier.xml [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+                <param name="homologies" type="select" multiple="true" optional="false" display="checkboxes" label="Pair Types to keep:">
+                    <option value="one-to-one" selected="true">one-to-one</option>
+                    <option value="one-to-many" selected="true">one-to-many</option>
+                    <option value="many-to-one" selected="true">many-to-one</option>
+                    <option value="many-to-many" selected="true">many-to-many</option>
+                    <option value="paralogs" selected="true">Paralogs</option>
+                </param>
+            </when>
+            <when value="csv">
+                <param name="homologies" type="select" multiple="true" optional="true" display="checkboxes" label="Discard families with:">
```
Comment (nsoranzo) [inline tools/ete/homology_classifier.xml]: Isn't it a bit confusing that this multiselect has the opposite meaning of the one above? Maybe the intended usage is not clear to me.

File: tools/ete/homology_classifier.py (hunk 10 above)
Comment (anilthanki) [reply tools/ete/homology_classifier.py]: This throws error "TypeError: list indices must be integers or slices, not tuple"

File: tools/ete/homology_classifier.xml (hunk 12 above)
Comment (anilthanki) [reply tools/ete/homology_classifier.xml]: We want to remove all the families which has any of these type of homology, in tabular form we just want to remove pair.

File: tools/ete/ete_homology_classifier.py [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+    # reads single gene tree
+    genetree = PhyloTree(options.genetree)
+
+    leaves_list = genetree.get_leaf_names()
+
+    species_list = []
+
+    for i, val in enumerate(leaves_list):
+        if '_' in leaves_list[i]:
+            species_list.append(leaves_list[i].split("_")[1])
```
Comment (nsoranzo) [inline tools/ete/ete_homology_classifier.py]: This for loop can be removed, lines 33-34 already do the same.

File: tools/ete/ete_homology_classifier.py [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+
+    homologies = {
+        'one-to-one': [],
+        'one-to-many': [],
+        'many-to-one': [],
+        'many-to-many': [],
+        'paralogs': []
+    }
+
+    # stores relevant homolgy types in dict
```
Comment (nsoranzo) [inline tools/ete/ete_homology_classifier.py]: s/homolgy/homology/

File: tools/ete/homology_classifier.py (hunk 10 above)
Comment (nsoranzo) [reply tools/ete/homology_classifier.py]: Ah, that should have been: `for leaf2 in leaves_list[i + 1:]:`

File: tools/ete/homology_classifier.xml (hunk 12 above)
Comment (nsoranzo) [reply tools/ete/homology_classifier.xml]: I still don't get why for tabular you ask which type to keep and for csv which type to discard.

**Comment by nsoranzo** [comment]
Can you add/change tests with a genetree that has more than one type of homologies?

File: tools/ete/homology_classifier.xml (hunk 12 above)
Comment (anilthanki) [reply tools/ete/homology_classifier.xml]: In tabular format families are shown as pairs, so we are able to filter on pairs, but in CSV format there are no pairs but user may want to filter out whole family where there is any of the selected homology type present

**Pull request merged**